allocations/tasks.py

```diff
@@ -0,0 +1,92 @@
+
+from datetime import datetime
+import logging
+import pytz
+
+from celery import task
+from django.db import transaction
+from allocations.models import Allocation
+from allocations.allocations_api import BalanceServiceClient
+
+logger = logging.getLogger(__name__)
+
+@task
+def activate_expire_allocations():
+    now = datetime.now(pytz.utc)
+    balance_service = BalanceServiceClient()
+    
+    # expire allocations
+    expired_allocations = Allocation.objects.filter(status='active', expiration_date__lte=now)
+    expired_alloc_count = 0
+    for alloc in expired_allocations:
+        # set status to inactive and set the final su_used in portal db
+        alloc.su_used = float(balance_service.call(alloc.project_charge_code)['used'])
+        alloc.status = 'inactive'
+        alloc.save()
+        # reset balance service
+        try:
+            balance_service.reset(alloc.project_charge_code)
+            expired_alloc_count = expired_alloc_count + 1
+        except Exception as be:
+            logger.exception(be)
+            # roll back portal db change
+            try: 
+                alloc.su_used = None
+                alloc.status = 'active'
+                alloc.save()
+                logger.info('successfully roll back portal db change for {}'.format(alloc.project_charge_code))
+            except Exception as re:
+                logger.exception(re)
+
+    logger.debug('need to expire {} allocations, and {} were actually expired'.format(len(expired_allocations), expired_alloc_count))
+    
+    # check projects with multiple active allocations
+    for proj in Allocation.objects.order_by().values_list('project_charge_code', flat=True).distinct():
+        project_active_allocations = list(Allocation.objects.filter(status='active', project_charge_code=proj))
+        if len(project_active_allocations) > 1:
+            logger.warning('project {} has more than one active allocations'.format(proj))
+            max_expiration_date = max([a.expiration_date for a in project_active_allocations])
+            is_active_alloc_set = False
+            for alloc in project_active_allocations:
+                if not is_active_alloc_set and alloc.expiration_date == max_expiration_date:
+                    is_active_alloc_set = True
+                else:
+                    alloc.status = 'inactive'
+                    alloc.save()
+
+    # activate allocations
+    approved_allocations = Allocation.objects.filter(status='approved', start_date__lte=now)
+    activated_alloc_count = 0
+    for alloc in approved_allocations:
+        # deactivate active allocation for the project and set status to active
+        project_active_allocations = Allocation.objects.filter(status='active', project_charge_code=alloc.project_charge_code)
+        proj_alloc = None
+        if len(project_active_allocations) > 0:
+            proj_alloc = project_active_allocations[0]
+        with transaction.atomic():
```

Comment:
I like this approach. I think we can make it even safer, while keeping it quite simple:

```python
try:
    with transaction.atomic():
        su_used = float(balance_service.call(alloc.project_charge_code)['used'])
        if proj_alloc:
            proj_alloc.su_used = su_used
            proj_alloc.status = 'inactive'
            proj_alloc.save()
        alloc.status = 'active'
        alloc.save()
        balance_service.recharge(alloc.project_charge_code, alloc.su_allocated)
except Exception:
    log.exception('Error activating project {}'.format(alloc.project_charge_code))
```

This should accomplish the following:
- if we fail to fetch the balance, nothing happens
- if one of the DB transactions fails, the transaction rolls back
- if we fail to recharge the allocation, the transaction rolls back
- if any allocation update fails, the system still proceeds to the next one in the list for processing (it doesn't abort)

I think this should ensure that we keep the two systems in sync, if I'm reasoning about it correctly.